#!/usr/bin/env cwl-runner
cwlVersion: v1.0
class: CommandLineTool


requirements:
- class: InlineJavascriptRequirement
- class: DockerRequirement
  dockerPull: biowardrobe2/bismark:v0.0.2


inputs:

  alignment_report:
    type: File
    inputBinding:
      prefix: "--alignment_report"
    label: "Bismark alignment and methylation report"
    doc: "Bismark generated alignment and methylation summary report"

  splitting_report:
    type: File?
    inputBinding:
      prefix: "--splitting_report"
    label: "Methylation extraction log"
    doc: "Log file giving summary statistics about methylation extraction"

  mbias_report:
    type: File?
    inputBinding:
      prefix: "--mbias_report"
    label: "Methylation bias plot"
    doc: "QC data showing methylation bias across read lengths"

  deduplication_report:
    type: File?
    inputBinding:
      prefix: "--dedup_report"
    label: "Bismark deduplication report"
    doc: "Bismark generated deduplication report"

  nucleotide_report:
    type: File?
    inputBinding:
      prefix: "--nucleotide_report"
    label: "Nucleotide coverage report"
    doc: "Bismark nucleotide coverage report"


outputs:

  collected_report:
    type: File
    label: "HTML report page"
    doc: "Bismark generated graphical HTML report page"
    outputBinding:
      glob: "*"


baseCommand: ["bismark2report"]


$namespaces:
  s: http://schema.org/

$schemas:
- http://schema.org/version/latest/schema.rdf

s:mainEntity:
  $import: ./metadata/bismark_metadata.yaml

s:name: "bismark_report"
s:downloadUrl: https://raw.githubusercontent.com/common-workflow-library/bio-cwl-tools/release/tools/bismark/bismark_report.cwl
s:codeRepository: https://github.com/common-workflow-library/bio-cwl-tools
s:license: http://www.apache.org/licenses/LICENSE-2.0

s:isPartOf:
  class: s:CreativeWork
  s:name: Common Workflow Language
  s:url: http://commonwl.org/

s:creator:
- class: s:Organization
  s:legalName: "Cincinnati Children's Hospital Medical Center"
  s:location:
  - class: s:PostalAddress
    s:addressCountry: "USA"
    s:addressLocality: "Cincinnati"
    s:addressRegion: "OH"
    s:postalCode: "45229"
    s:streetAddress: "3333 Burnet Ave"
    s:telephone: "+1(513)636-4200"
  s:logo: "https://www.cincinnatichildrens.org/-/media/cincinnati%20childrens/global%20shared/childrens-logo-new.png"
  s:department:
  - class: s:Organization
    s:legalName: "Allergy and Immunology"
    s:department:
    - class: s:Organization
      s:legalName: "Barski Research Lab"
      s:member:
      - class: s:Person
        s:name: Michael Kotliar
        s:email: mailto:misha.kotliar@gmail.com
        s:sameAs:
        - id: http://orcid.org/0000-0002-6486-3898

doc: |
  This tool uses a Bismark alignment report to generate a graphical HTML report page. Optionally, further reports of
  the Bismark suite such as deduplication, methylation extractor splitting or M-bias reports can be specified as well.

  Skipped arguments:
    -o/--output
    --dir

s:about: |
  USAGE: bismark2report [options]

  -o/--output <filename>     Name of the output file (optional). If not specified explicitly, the output filename will be derived
                            from the Bismark alignment report file. Specifying an output filename only works if the HTML report is
                            to be generated for a single Bismark alignment report (and potentially additional reports).

  --dir                      Output directory. Output is written to the current directory if not specified explicitly.


  --alignment_report FILE    If not specified explicitly, bismark2report attempts to find Bismark report file(s) in the current
                            directory and produces a separate HTML report for each mapping report file. Based on the basename of
                            the Bismark mapping report, bismark2report will also attempt to find the other Bismark reports (see below)
                            for inclusion into the HTML report. Specifying a Bismark alignment report file is mandatory.

  --dedup_report FILE        If not specified explicitly, bismark2report attempts to find a deduplication report file with the same
                            basename as the Bismark mapping report (generated by deduplicate_bismark) in the
                            current working directory. Including a deduplication report is optional, and using the FILE 'none'
                            will skip this step entirely.

  --splitting_report FILE    If not specified explicitly, bismark2report attempts to find a splitting report file with the same
                            basename as the Bismark mapping report (generated by the Bismark methylation extractor) in the current
                            working directory. Including a splitting report is optional, and using the FILE 'none' will skip this
                            step entirely.

  --mbias_report FILE        If not specified explicitly, bismark2report attempts to find a single M-bias report file with the same
                            basename as the Bismark mapping report (generated by the Bismark methylation extractor) in the current
                            working directory. Including an M-Bias report is optional, and using the FILE 'none' will skip this step
                            entirely.

  --nucleotide_report FILE   If not specified explicitly, bismark2report attempts to find a single nucleotide coverage report file
                            with the same basename as the Bismark mapping report (generated by Bismark with the option
                            '--nucleotide_coverage') in the current working directory. Including a nucleotide coverage statistics
                            report is optional, and using the FILE 'none' will skip this report entirely.

                            Script last modified: 07 August 2018
